core/network-libp2p/src/custom_proto/handler.rs

```diff
@@ -165,44 +158,199 @@ struct PerProtocolNormalState<TMessage, TSubstream> {
+				Ok(Async::Ready(Some(RegisteredProtocolEvent::Message(message)))) => {
+					if message.request_id() == CustomMessageId::Response(request_id) {
+						let event = CustomProtoHandlerOut::CustomMessage {
+							message
+						};
+						self.shutdown.push(substream);
+						return Some(event);
+					} else {
+						self.shutdown.push(substream);
+						let event = CustomProtoHandlerOut::ProtocolError {
+							is_severe: true,
+							error: format!("Request ID doesn't match substream: expected {:?}, \
+								got {:?}", request_id, message.request_id()).into(),
+						};
+						return Some(event);
+					}
+				},
+				Ok(Async::Ready(Some(RegisteredProtocolEvent::Clogged { .. }))) =>
+					unreachable!("Cannot receive Clogged message with new protocol version; QED"),
+				Ok(Async::NotReady) =>
+					self.pending_response.push((request_id, substream)),
+				Ok(Async::Ready(None)) => {
+					self.shutdown.push(substream);
+					let event = CustomProtoHandlerOut::ProtocolError {
+						is_severe: false,
+						error: format!("Request ID {:?} didn't receive an answer", request_id).into(),
+					};
+					return Some(event);
+				}
+				Err(err) => {
+					self.shutdown.push(substream);
+					let event = CustomProtoHandlerOut::ProtocolError {
+						is_severe: false,
+						error: format!("Error while waiting for an answer for {:?}: {}",
+							request_id, err).into(),
+					};
+					return Some(event);
+				}
+			}
+		}
 
-		let return_value;
+		for n in (0..self.incoming_substreams.len()).rev() {
+			let mut substream = self.incoming_substreams.swap_remove(n);
+			match substream.poll() {
+				Ok(Async::Ready(Some(RegisteredProtocolEvent::Message(message)))) => {
+					if let CustomMessageId::Request(id) = message.request_id() {
+						self.pending_send_back.push((id, substream));
+						return Some(CustomProtoHandlerOut::CustomMessage {
+							message
+						});
+					} else if let CustomMessageId::OneWay = message.request_id() {
+						self.shutdown.push(substream);
+						return Some(CustomProtoHandlerOut::CustomMessage {
+							message
+						});
+					} else {
+						self.shutdown.push(substream);
+						return Some(CustomProtoHandlerOut::ProtocolError {
+							is_severe: true,
+							error: format!("Received response in new substream").into(),
+						});
+					}
+				},
+				Ok(Async::Ready(Some(RegisteredProtocolEvent::Clogged { .. }))) =>
+					unreachable!("Cannot receive Clogged message with new protocol version; QED"),
+				Ok(Async::NotReady) =>
+					self.incoming_substreams.push(substream),
+				Ok(Async::Ready(None)) => {}
+				Err(err) => {
+					self.shutdown.push(substream);
+					return Some(CustomProtoHandlerOut::ProtocolError {
+						is_severe: false,
+						error: format!("Error in incoming substream: {}", err).into(),
+					});
+				}
+			}
+		}
 
-		self.state = match mem::replace(&mut self.state, PerProtocolState::Poisoned) {
-			PerProtocolState::Poisoned => {
+		shutdown_list(&mut self.shutdown);
+		None
+	}
+}
```

Comment:
Just noticed that entire code was just moved, so feel free to disregards the comments.